koku/kafka_utils/utils.py

```diff
@@ -66,10 +67,14 @@ def _get_producer_config(address, conf_settings):  # pragma: no cover
 
 def get_producer(conf_settings=None, address=Config.INSIGHTS_KAFKA_ADDRESS):  # pragma: no cover
     """Create a Kafka producer."""
+    global PRODUCER
+    if PRODUCER:
+        return PRODUCER
     if conf_settings is None:
         conf_settings = {}
     conf = _get_producer_config(address, conf_settings)
-    return Producer(conf)
+    PRODUCER = Producer(conf)
+    return PRODUCER
```

Comment:
This looks like we want `Producer` to be a singleton. Rather than use a global, make a class that returns a singleton.

Reply:
Here's an example Singleton metaclass:
```python
class SingletonMeta(type):
    """Metaclass for classes that wish to implement Singleton functionality.
    If an instance of a class already exists, it is returned. Otherwise a new
    instance is created and returned in a thread safe manner.
    """

    _instance = None
    _rlock = RLock()

    def __call__(cls, *args, **kwargs):
        if cls._instance is not None:
            return cls._instance

        with cls._rlock:
            if cls._instance is None:
                print("Creating with lock")
                cls._instance = super().__call__(*args, **kwargs)

        return cls._instance


class ProducerSingleton(Producer, metaclass=SingletonMeta):
    ...
```

Then `get_producer` becomes much simpler.

```python
def get_producer(conf_settings=None, address=Config.INSIGHTS_KAFKA_ADDRESS):  # pragma: no cover
    """Create a Kafka producer."""
    if conf_settings is None:
        conf_settings = {}
    conf = _get_producer_config(address, conf_settings)
    return ProducerSingleton(conf)
```

Reply:
Example of this in action:

```shell
# Using Singleton meta class

In [3]: id(get_producer())
Out[3]: 4451232640

In [4]: id(get_producer())
Out[4]: 4451232640

In [5]: id(get_producer())
Out[5]: 4451232640

In [6]: id(get_producer())
Out[6]: 4451232640
```

```shell
# Behavior without changes in this PR

In [2]: id(get_producer())
Out[2]: 4426746336

In [3]: id(get_producer())
Out[3]: 4426821120

In [4]: id(get_producer())
Out[4]: 4426775888
```